stdlib/public/core/NormalizedCodeUnitIterator.swift

```diff
@@ -348,84 +429,76 @@ struct _NormalizedUTF16CodeUnitIterator: IteratorProtocol {
         }
-
-        guard let count = normalizationBuffer!.withUnsafeMutableBufferPointer({
-          (normalizationBufferPtr) -> Int? in
-          guard let filled = source.tryFill(into: normalizationBufferPtr)
-          else {
-            fatalError("Invariant broken, buffer should have space")
+        return filled
+      }
+    }
+  }
+  
+  private mutating func normalizeIntoSegmentBuffer() {
+    do {
+      try intermediateBuffer.withUnsafeBufferPointer { intermediateBuff in
+        try segmentBuffer.withUnsafeMutableBufferPointer { segmentBuff in
+          if let filled = _tryNormalize(intermediateBuff, into: segmentBuff) {
+            return filled
+          } else {
+            throw NormalizationError.bufferTooSmall
           }
-          return overflowBuffer!.withUnsafeMutableBufferPointer {
-            (overflowBufferPtr) -> Int? in
-            return _tryNormalize(
-              UnsafeBufferPointer(rebasing: normalizationBufferPtr[..<filled]),
-              into: overflowBufferPtr
-            )
+        }
+      }
+    } catch {
+      segmentBuffer.promote(source.remaining + intermediateBuffer.count, multiplier: _Normalization._maxNFCExpansionFactor)
+      intermediateBuffer.withUnsafeBufferPointer { intermediateBuff in
+        segmentBuffer.withUnsafeMutableBufferPointer { segmentBuff in
+          guard let filled = _tryNormalize(intermediateBuff, into: segmentBuff) else {
+            fatalError("Heap buffer wasn't big enough")
           }
-        }) else {
-          fatalError("Invariant broken, overflow buffer should have space")
+          return filled
         }
-
-        overflowBufferCount = count
       }
     }
-
-    //exactly one of the buffers should have code units for us to return
-    _sanityCheck((segmentBufferCount == 0)
-              != ((overflowBuffer?.count ?? 0) == 0))
-
-    if segmentBufferIndex < segmentBufferCount {
-      let index = segmentBufferIndex
-      segmentBufferIndex += 1
-      return segmentBuffer[index]
-    } else if overflowBufferIndex < overflowBufferCount {
-      _sanityCheck(overflowBufferIndex < overflowBuffer!.count)
-      let index = overflowBufferIndex
-      overflowBufferIndex += 1
-      return overflowBuffer![index]
+  }
+  
+  mutating func next() -> CodeUnit? {
+    if !segmentBuffer.canRead {
+      segmentBuffer.reset()
+    }
+    
+    if source.remaining <= 0 && !segmentBuffer.canRead {
+      // Our source of code units to normalize is empty and our buffer from 
+      // previous normalizations is also empty.
+      return nil
+    }
+    
+    if !segmentBuffer.canRead {
+      //time to fill the buffer if possible. Otherwise we are done, return nil
+      // Normalize segment, and then compare first code unit
+      fillIntermediateBuffer()
+      normalizeIntoSegmentBuffer()
+    }
+    
+    if segmentBuffer.canRead {
+      return segmentBuffer.read()
     } else {
-        return nil
+      return nil
```

Comment:
The logic here can be simplified. If there's nothing left to read from the source or our buffer, we return nil. Otherwise, there shouldn't be a path to returning nil.